src/trace/persistent/cursor.rs

```diff
@@ -0,0 +1,284 @@
+use super::{ReusableEncodeBuffer, Values, BINCODE_CONFIG, ROCKS_DB_INSTANCE};
+use crate::algebra::{MonoidValue, PartialOrder};
+use crate::trace::{Batch, BatchReader, Cursor};
+
+/// The cursor for the persistent trace.
+pub struct PersistentTraceCursor<'s, B: Batch + 's>
+where
+    B::Val: Encode + Decode,
+    B::Time: Encode + Decode,
+    B::R: MonoidValue + Encode + Decode,
+    <B as BatchReader>::Val: Ord,
+{
+    db_iter: DBRawIterator<'s>,
+
+    cur_key: Option<B::Key>,
+    cur_vals: Option<Values<B::Val, B::Time, B::R>>,
+    /// Current index of iterator in `cur_vals`.
+    ///
+    /// Value will be `0 <= val_idx < cur_vals.len()`.
+    val_idx: usize,
+
+    cursor_upper_bound: Option<B::Key>,
+    /// Temporary storage serializing seeked keys.
+    tmp_key: ReusableEncodeBuffer,
+}
+
+impl<'s, B> PersistentTraceCursor<'s, B>
+where
+    B: Batch + 's,
+    B::Key: Ord + Clone + Encode + Decode,
+    B::Val: Ord + Encode + Decode,
+    B::Time: Encode + Decode,
+    B::R: MonoidValue + Encode + Decode,
+{
+    /// Loads the current key and its values from RocksDB and stores them in the
+    /// [`PersistentTraceCursor`] struct.
+    ///
+    /// # Panics
+    /// - In case the `db_iter` is invalid.
+    fn update_current_key_weight(&mut self) {
+        assert!(self.db_iter.valid());
+        let (key, values) = PersistentTraceCursor::<'s, B>::read_key_val_weights(&mut self.db_iter);
+
+        self.val_idx = 0;
+        self.cur_key = key;
+        self.cur_vals = values;
+    }
+
+    /// Loads the current key and its values from RocksDB.
+    ///
+    /// # Returns
+    /// - The key and its values.
+    /// - `None` if the iterator is invalid.
+    #[allow(clippy::type_complexity)]
+    fn read_key_val_weights(
+        iter: &mut DBRawIterator<'s>,
+    ) -> (Option<B::Key>, Option<Values<B::Val, B::Time, B::R>>) {
+        loop {
+            if !iter.valid() {
+                return (None, None);
+            }
+
+            if let (Some(k), Some(v)) = (iter.key(), iter.value()) {
+                let (key, _) =
+                    decode_from_slice(k, BINCODE_CONFIG).expect("Can't decode current key");
+                let (values, _): (PersistedValue<B::Val, B::Time, B::R>, usize) =
+                    decode_from_slice(v, BINCODE_CONFIG).expect("Can't decode current value");
+                match values {
+                    PersistedValue::Values(vals) => {
+                        return (Some(key), Some(vals));
+                    }
+                    PersistedValue::Tombstone => {
+                        // Skip tombstones:
+                        // they will be deleted by the compaction filter at a later point in time
+                        iter.next();
+                        continue;
+                    }
+                }
+            } else {
+                // This holds according to the RocksDB C++ API docs
+                unreachable!("db_iter.valid() implies Some(key)")
```

Comment:
Not sure I understand because unreachable leads to `panic!` too (as opposed to `unreachable_unchecked!`)?

From https://doc.rust-lang.org/std/macro.unreachable.html:
> This will always [panic!](https://doc.rust-lang.org/core/macro.panic.html) because unreachable! is just a shorthand for panic! with a fixed, specific message.

Reply:
I didn't realize this was the case.